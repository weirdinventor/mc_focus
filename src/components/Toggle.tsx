import React from 'react';

const componentStyles = `
  .toggle-button {
    background: none;
    border: none;
    padding: 0;
    cursor: pointer;
    position: relative;
    display: flex;
    align-items: center;
    overflow: hidden;
    width: var(--container-width);
    height: var(--container-height);
    border-radius: var(--container-height);
    background-color: var(--off-background);
  }

  .toggle-inner-bg {
    position: absolute;
    width: 100%;
    height: 100%;
    background-color: var(--on-background);
    opacity: 0;
    transition: opacity 0.3s ease-in-out;
  }

  .toggle-knob {
    position: absolute;
    background-color: #FFFFFF;
    width: var(--knob-size);
    height: var(--knob-size);
    border-radius: 50%;
    left: var(--padding);
    transition: left 0.3s cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }

  .toggle-button.activated .toggle-inner-bg {
    opacity: 1;
  }

  .toggle-button.activated .toggle-knob {
    left: calc(var(--container-width) - var(--knob-size) - var(--padding));
  }
`;

interface ToggleProps {
  activated: boolean;
  onPressHandler: () => void;
  size?: number;
}

export const Toggle = ({
  activated,
  onPressHandler,
  size = 32,
}: ToggleProps) => {
  const offBackground = '#333333';
  const onBackground = '#EEEEEE';

  const KNOB_SIZE = size * 0.8;
  const CONTAINER_WIDTH = size * 1.8;
  const PADDING = (size - KNOB_SIZE) / 2;

  const buttonClassName = `toggle-button ${activated ? 'activated' : ''}`;

  const buttonStyle = {
    '--container-width': `${CONTAINER_WIDTH}px`,
    '--container-height': `${size}px`,
    '--knob-size': `${KNOB_SIZE}px`,
    '--padding': `${PADDING}px`,
    '--off-background': offBackground,
    '--on-background': onBackground,
  };

  return (
    <>
      <style>{componentStyles}</style>
      <button
        role="switch"
        aria-checked={activated}
        onClick={onPressHandler}
        className={buttonClassName}
        // --- LA CORRECTION EST ICI ---
        // On dit Ã  TypeScript de traiter cet objet comme un style CSS valide.
        style={buttonStyle as React.CSSProperties}
      >
        <div className="toggle-inner-bg" />
        <div className="toggle-knob" />
      </button>
    </>
  );
};