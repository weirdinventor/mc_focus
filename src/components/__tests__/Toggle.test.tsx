import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Toggle } from '@components/Toggle';

describe('[COMPONENTS]: Toggle', () => {
  const mockOnPressHandler = jest.fn();

  it('renders correctly', () => {
    const tree = render(
      <Toggle activated={false} onPressHandler={mockOnPressHandler} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with active state', () => {
    const tree = render(
      <Toggle activated={true} onPressHandler={mockOnPressHandler} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls onPressHandler when pressed', () => {
    const { getByRole } = render(
      <Toggle activated={false} onPressHandler={mockOnPressHandler} />,
    );
    const button = getByRole('togglebutton');
    fireEvent.press(button);
    expect(mockOnPressHandler).toHaveBeenCalled();
  });
});
