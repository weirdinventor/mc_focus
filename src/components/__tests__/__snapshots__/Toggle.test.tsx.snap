// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[COMPONENTS]: Toggle renders correctly 1`] = `
<View
  accessibilityRole="togglebutton"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 16,
          "borderWidth": 0,
          "flexDirection": "row",
          "flexGrow": 0,
          "flexShrink": 0,
          "justifyContent": "space-between",
          "overflow": "hidden",
        },
        {
          "backgroundColor": undefined,
          "height": 32,
          "width": 57.6,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "borderColor": undefined,
            "height": "100%",
            "overflow": "hidden",
            "paddingEnd": 4,
            "paddingStart": 4,
            "position": "absolute",
            "width": "100%",
          },
          {
            "backgroundColor": undefined,
          },
          {
            "opacity": 0,
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 12,
            "height": 26,
            "position": "absolute",
            "width": 26,
          },
          {
            "marginStart": 4,
            "start": "0%",
          },
          {
            "height": 25.6,
            "width": 25.6,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`[COMPONENTS]: Toggle renders correctly with active state 1`] = `
<View
  accessibilityRole="togglebutton"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 16,
          "borderWidth": 0,
          "flexDirection": "row",
          "flexGrow": 0,
          "flexShrink": 0,
          "justifyContent": "space-between",
          "overflow": "hidden",
        },
        {
          "backgroundColor": undefined,
          "height": 32,
          "width": 57.6,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "borderColor": undefined,
            "height": "100%",
            "overflow": "hidden",
            "paddingEnd": 4,
            "paddingStart": 4,
            "position": "absolute",
            "width": "100%",
          },
          {
            "backgroundColor": undefined,
          },
          {
            "opacity": 1,
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 12,
            "height": 26,
            "position": "absolute",
            "width": 26,
          },
          {
            "marginStart": -29.6,
            "start": "100%",
          },
          {
            "height": 25.6,
            "width": 25.6,
          },
        ]
      }
    />
  </View>
</View>
`;
